### PR TITLE
Allow explicit cast between varlist CAST(var, varlist type)

### PR DESCRIPTION
**Pull request by AEsir777**
Benchmark: (size of list between 0 to 40, select condition size(list) > 25)
| 1000000 x Int128[][] | cast whole list | selected vector |
| --------| -------| ------|
| | 1745.701000| 3619 before malloc() |
| | 937.575| |
|100000 x INT128[] to INT32[] (list of ints between 1 to 1000 and select condition size(list) > 800)|7776.816 | 7986.32|
|> 950|2106 | 2151.219|
|all | 15766| 21384| 

ToDo:
- ~add more tests for list of nested types~
- ~add cast_error cases~

### CODE REVIEW

File: src/function/vector_cast_functions.cpp [hunk 1]
```diff
@@ -411,12 +526,30 @@ static std::unique_ptr<ScalarFunction> bindCastToNumericFunction(
...
+template<typename /*DST_TYPE*/, typename EXECUTOR = UnaryFunctionExecutor>
+static std::unique_ptr<ScalarFunction> bindCastBetweenNested(
+    const std::string& functionName, LogicalTypeID sourceTypeID, LogicalTypeID targetTypeID) {
+    scalar_exec_func func;
+    switch (sourceTypeID) {
+    case LogicalTypeID::VAR_LIST: {
+        func = varListCastExecFunction<EXECUTOR>;
+    } break;
+    default:
+        throw ConversionException{stringFormat("Unsupported casting function from {} to {}.",
```
Comment (acquamarin) [inline src/function/vector_cast_functions.cpp:538]: Use //lcov_excl_start

File: src/include/function/unary_function_executor.h [hunk 2]
```diff
@@ -89,6 +89,28 @@ struct UnaryUDFFunctionWrapper {
...
+struct CastChildFunctionExecutor {
+    template<typename OPERAND_TYPE, typename RESULT_TYPE, typename FUNC, typename OP_WRAPPER>
+    static void executeSwitch(
+        common::ValueVector& operand, common::ValueVector& result, void* dataPtr) {
+        // this vector is of var list type and the child vector is of non-nested types then cast
+        KU_ASSERT(operand.dataType.getLogicalTypeID() == common::LogicalTypeID::VAR_LIST &&
+                  result.dataType.getLogicalTypeID() == common::LogicalTypeID::VAR_LIST);
+        result.resetAuxiliaryBuffer();
+        auto childNum = common::ListVector::getDataVectorSize(&operand);
+        auto inputChildVector = (common::ListVector::getDataVector(&operand));
```
Comment (acquamarin) [inline src/include/function/unary_function_executor.h:101]: remove the redundant '()'

File: src/include/function/unary_function_executor.h [hunk 3]
```diff
@@ -89,6 +89,28 @@ struct UnaryUDFFunctionWrapper {
     }
 };
 
+struct CastChildFunctionExecutor {
+    template<typename OPERAND_TYPE, typename RESULT_TYPE, typename FUNC, typename OP_WRAPPER>
+    static void executeSwitch(
+        common::ValueVector& operand, common::ValueVector& result, void* dataPtr) {
+        // this vector is of var list type and the child vector is of non-nested types then cast
+        KU_ASSERT(operand.dataType.getLogicalTypeID() == common::LogicalTypeID::VAR_LIST &&
```
Comment (acquamarin) [inline src/include/function/unary_function_executor.h:97]: This executor looks incorrect to me:
1. The executor doesn't handle nulls correctly. We should loop through each value in the operand and set null in the result accordingly.
2. If the operand state is selected, we should skip casting unselected elements.

File: src/function/vector_cast_functions.cpp [hunk 4]
```diff
@@ -86,6 +103,78 @@ static void StringtoFixedListCastExecFunction(
...
+        resultChildVector->setNull(i, inputChildVector->isNull(i));
+        if (!resultChildVector->isNull(i)) {
+            CastString::castToFixedList(
+                inputChildVector->getValue<ku_string_t>(i), resultChildVector, i, csvReaderConfig);
+        }
+    }
+}
+
+template<typename /*EXECUTOR*/ = UnaryFunctionExecutor>
+static void varListCastExecFunction(
```
Comment (acquamarin) [inline src/function/vector_cast_functions.cpp:128]: As we discussed, we don't need to template this function

**Review by acquamarin — CHANGES_REQUESTED** [review]

File: src/include/function/unary_function_executor.h [hunk 5]
```diff
@@ -89,6 +89,27 @@ struct UnaryUDFFunctionWrapper {
...
+                  result.dataType.getLogicalTypeID() == common::LogicalTypeID::VAR_LIST);
+        auto childNum = common::ListVector::getDataVectorSize(&operand);
+        auto inputChildVector = common::ListVector::getDataVector(&operand);
+        auto resultChildVector = (common::ListVector::getDataVector(&result));
+        for (auto i = 0u; i < childNum; i++) {
+            resultChildVector->setNull(i, inputChildVector->isNull(i));
+            if (!resultChildVector->isNull(i)) {
+                // cast position i in child data vector
+                OP_WRAPPER::template operation<OPERAND_TYPE, RESULT_TYPE, FUNC>(
```
Comment (AEsir777) [inline src/include/function/unary_function_executor.h:106]: This is covered

**Review by acquamarin — APPROVED** [review]

File: src/function/vector_cast_functions.cpp [hunk 6]
```diff
@@ -86,6 +103,67 @@ static void StringtoFixedListCastExecFunction(
...
+    auto resultVector = &result;
+
+    auto childNum = ListVector::getDataVectorSize(inputVector.get());
+    ListVector::addList(&result, childNum);
+    auto listEntrySize = inputVector->state->selVector
+                             ->selectedPositions[inputVector->state->selVector->selectedSize - 1] +
+                         1;
+    memcpy(resultVector->getData(), inputVector->getData(),
+        listEntrySize * resultVector->getNumBytesPerValue());
+    resultVector->setNullFromBits(inputVector->getNullMaskData(), 0, 0, listEntrySize);
```
[the remark this answers is not in the document]
Comment (AEsir777) [reply src/function/vector_cast_functions.cpp:141]: This is a private function for NullMask. `setNullFromBits` is only public method, which calls `setNullFromBits`.

**Pull request merged**